### ISSUE TITLE
Question: how to make the uploaded image public after uploading

### ISSUE THREAD

**Issue by MaeseppTarvo** [Superbalist/laravel-google-cloud-storage#21]
Hey, I have a question. I am trying to upload image and make it public. I've made the right permissions to my bucket but for some reason I am getting exception:

```
{
"error": {
"errors": [
{
"domain": "global",
"reason": "insufficientPermissions",
"message": "Insufficient Permission"
}
],
"code": 403,
"message": "Insufficient Permission"
}
}
```

This is how I upload and it gives me the above error:

```$i = 1;
foreach($request->file('post_image') as $image){
               $filename = $post->id.'_'.$i.'.jpg';
               $location = 'posts/'.$post->id;
               $storage = Storage::disk('gcs')->setVisibility($filename, 'public')->put($location, $image);
               $i++;
}
```

But it works like this:
```$i = 1;
foreach($request->file('post_image') as $image){
               $filename = $post->id.'_'.$i.'.jpg';
               $location = 'posts/'.$post->id;
               $storage = Storage::disk('gcs')->put($location, $image);
               $i++;
}
```

**Comment by zoidbergwill** [comment]
And if you try set visibility afterwards?

**Comment by MaeseppTarvo** [comment]
@zoidbergwill If I do like that:

```
$storage = Storage::disk('gcs')->put($location, $image);
$storage->setVisibility($filename, 'public');
```

I get 

> Call to a member function setVisibility() on string

**Comment by zoidbergwill** [comment]
And what about the following?

```
Storage::disk('gcs')->setVisibility($location, 'public');
```

**Comment by zoidbergwill** [comment]
This looks like a dupe of https://github.com/Superbalist/laravel-google-cloud-storage/issues/8, I think.

We should improve the docs, and errors if possible.

**Comment by MaeseppTarvo** [comment]
@zoidbergwill yep it is pretty much the same, however I am getting errors while setting the visibility. If I do as you said then I get the 

> Insufficient Permission 

error

**Comment by zoidbergwill** [comment]
What permissions does the service account you're using have?

**Comment by MaeseppTarvo** [comment]
Those are the permissions:

<img width="352" alt="screen shot 2017-11-30 at 17 18 52" src="https://user-images.githubusercontent.com/8416295/33438118-951259ac-d5f2-11e7-93cc-065cc89decb8.png">

**Comment by zoidbergwill** [comment]
Weird. Not sure how those match to either of these:

https://cloud.google.com/storage/docs/access-control/iam-roles
https://cloud.google.com/storage/docs/access-control/iam-permissions

Ahhhhh, the following page mentions that you need Storage Owner for editing bucket metadata and ACLs:

https://cloud.google.com/storage/docs/access-control/lists#permissions

I assume it means bucket and object ACLs.

**Comment by zoidbergwill** [comment]
Another option would be using a bucket with a default object ACL of public?

**Comment by MaeseppTarvo** [comment]
I tried the following: `gsutil iam ch allUsers:objectViewer gs://ex-bucket` but it still gives me the same error.

**Comment by qasimalbaqali** [comment]
This worked for me

```
$images_path = 'pool_images/' . $pool->id . '/';
            $storage = Storage::disk('gcs');

            $image = $request->file('image');
            $image_new_name = str_replace(' ', '_', $image->getClientOriginalName());

            $storage->putFileAs($images_path, $image, $image_new_name, 'public');
```

Give it a try

**Comment by ultrono** [comment]
Just in case anyone else comes here from Google, I was having a similar issue. I uploaded files without issue and set the visibility to public. However, when viewing the public url I was prompted to login to my Google account.

This was caused by me using an incorrect base url. I was using https://storage.cloud.google.com/bucket/file.txt instead of https://storage.googleapis.com/bucket/file.txt. using the latter url resolved my issues.
